### DISCUSSION TITLE
[5.x]: Address translation

### DISCUSSION THREAD

**Discussion by olsp** [craftcms/cms#15699] [Ideas]
### What happened?

I would like to translate the subfields of an address field. The city for example has different names in different languages but also the street name may be different. I didn't find a way to translate those fields, is this currently possible?

### Craft CMS version

5.4.1

**Comment by i-just** [comment]
Hi, thanks for getting in touch. Can you please clarify what you'd like to do? Do you want to translate the labels of the address subfield; translate the predefined values that show in dropdowns (like in the country field); something else?

**Comment by olsp** [comment]
Sorry for not being more precise in the first place.
What I would like to do is to translate the contents of the subfields.

So for example I have an address, where the city is Vienna (in English). On the German site, it should be called Wien.
Right now, only the label field can be translated, as indicated with the icon next to the label.

<img width="445" alt="address-fields" src="https://github.com/user-attachments/assets/79e91180-cbd6-4972-a03a-047d5aeefe77">

**Reply by brandonkelly** [reply]
Addresses are currently not localizable, so I just converted your issue into a discussion. We’ll update this if that ends up changing.

**Reply by brandonkelly** [reply]
Craft 5.4.3 is out with a fix to stop showing that translatable icon on the Label field.

**Comment by olsp** [comment]
One more thing to add here: the field handles are also not shown when editing an address.

**Reply by brandonkelly** [reply]
Please post a new issue with that, thanks!

**Comment by svondervoort** [comment]
We are currently running into the same issue. We have a multi-language Craft and an address field on an entry which can not be translated, but depending on the site/language the address should be different since the head office is located somewhere else.

How can we achieve this in the current setup using the build in Address field?

**Reply by brandonkelly** [reply]
You could solve it using a Matrix field:

1. Create an “Address” entry type with just the Addresses field (no title, etc.)
2. Create an “Address” Matrix field with the following settings:
   - Entry types: just “Address”
   - Propagation method: “Only save entries to the site they were created in”
   - Min/Max Entries: `1`
   - View mode: “As inline-editable blocks”
3. Assign the “Address” Matrix field to whatever entry type currently has the Addresses field.

To migrate the existing address data into the nested Addresses field, you could use the “Copy address” action within the old Addresses field instance to copy existing addresses into new instance nested within your Matrix field, for each entry.

https://github.com/user-attachments/assets/d63d55bc-be33-4f06-9fb7-e4a35998d2be

If there’s too many entries/addresses to do that manually, you can create a [content migration](https://craftcms.com/docs/5.x/extend/migrations.html) with the following code in `safeUp()`:

```php
use craft\elements\Address;
use craft\elements\Entry;
use craft\helpers\Db;

$entries = Entry::find()
    ->section('mySection')
    ->type('myType')
    ->with([
        ['localized', ['status' => null]],
    ])
    ->status(null);

$addressMatrixField = Craft::$app->fields->getFieldByHandle('addressMatrix');

foreach (Db::each($entries) as $entry) {
    /** @var Entry $entry */
    /** @var Address|null $address */
    $address = $entry->address->one();
    if ($address) {
        /** @var Entry[] $allSiteEntries */
        $allSiteEntries = [
            $entry,
            ...$entry->getLocalized()->all(),
        ];

        foreach ($allSiteEntries as $siteEntry) {
            $addressEntry = $siteEntry->{$addressMatrixField->handle}->one();
            if (!$addressEntry) {
                $siteEntry->{$addressMatrixField->handle} = [
                    'new1' => [
                        'type' => 'address',
                        'enabled' => true,
                    ]
                ];
                Craft::$app->elements->saveElement($siteEntry);
                $addressEntry = $siteEntry->{$addressMatrixField->handle}->one();
            }
            Craft::$app->elements->duplicateElement($address, [
                'primaryOwnerId' => $addressEntry->id,
                'ownerId' => $addressEntry->id,
                'siteId' => $siteEntry->siteId,
            ]);
        }
    }
}
```

**Reply by svondervoort** [reply]
Ah yes thats what we ended up doing. putting it inside a matrix. didn't know about the copy feature on an address, thats a nice addition. Thanks for the tips.
